### ISSUE TITLE
Random failing ParallelBuildChainTest.testIndividualProjectBuilds_ProjectRelaxedRule

### ISSUE THREAD

**Issue by HeikoKlare** [eclipse-platform/eclipse.platform#657]
The `ParallelBuildChainTest.testIndividualProjectBuilds_ProjectRelaxedRule` is randomly failing: https://github.com/eclipse-platform/eclipse.platform/pull/655/checks?check_run_id=16473097978

```
java.lang.AssertionError: 
too few builds have run in parallel
Expected: a value equal to or greater than <3>
     but: <2> was less than <3>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.eclipse.core.tests.internal.builders.ParallelBuildChainTest.assertMinimumNumberOfSimultaneousBuilds(ParallelBuildChainTest.java:425)
	at org.eclipse.core.tests.internal.builders.ParallelBuildChainTest.lambda$1(ParallelBuildChainTest.java:140)
	at org.eclipse.core.tests.internal.builders.ParallelBuildChainTest.executeIndividualFullProjectBuilds(ParallelBuildChainTest.java:414)
	at org.eclipse.core.tests.internal.builders.ParallelBuildChainTest.testIndividualProjectBuilds_ProjectRelaxedRule(ParallelBuildChainTest.java:138)
```

The tests have recently been rewritten in response to recurring failures and inefficiency (see #609 and #641). The error logs show that one of the rewritten test cases fails because one of the builds that should run in parallel is actually not started in parallel but waits for another build to finish (test execution time is higher than 30 seconds, which is the execution time of a long running builds). I have no clue why the builds are not run in parallel yet. The jobs running the builds are started properly and the scheduling rules allow them to be executed in parallel. Still there is something that randomly blocks one at least one of the build operations. The error cannot be reproduced locally. 

I will at least improve the assertions, as the test should already fail because not all builds have started in time (and log which ones have not been executed). This does not happen because the timeout is higher than the execution time of a single build operation.

**Comment by iloveeclipse** [comment]
> Still there is something that randomly blocks one at least one of the build operations.

May be it's just small number of cores in the target environment, so even if nothing blocks the job, there is no free core to run it?

**Comment by HeikoKlare** [comment]
Thanks for your reply, @iloveeclipse! That enabled me to reproduce the problem locally. Physically, the number of available cores should be no limitation, as the build operations do not perform actual work but just yield the thread that executes them until they are explicitly terminated. So it could rather be a logical limitation concerning the number of available workers.
However, there seems to be an (implicit) limitation in the number of parallel project builds somewhere around the JobManager or the BuildManager that might be related to the execution environment: When increasing the number of projects that are built in parallel (while increasing the number of allowed parallel builds accordingly), the test starts to fail locally as well. There is no determinism with respect to the number of parallel builds resulting in test failures. On my machine it starts to fail randomly with 35 parallel builds and will rather constantly fail with more than 40 parallel builds (which is slightly higher than two times the number of available cores).

First debugging results show that builds become blocked by other build jobs although their scheduling rules are non-conflicting. And the wait operation for the blocking job to change its state will then never return.
I will analyze the problem now that I can reproduce it.

**Comment by HeikoKlare** [comment]
Should be fixed by #661.
